Docs: Add explanation for how per process power works

mrchrisadams opened the pull request:
Hi Benoit, 

Based on our discussions in #38 , I've written this up to add to the docs in mdbook.

I've left out the info about virtual machine scenarios, and multiple processor scenarios, as I agree with you that they can be address in separate pages, not least to give people who might be up for contributing to a firecracker version some clearer pointers.

### Unable to compile the docs locally though

I'm having a hard time compiling the mdbook docs locally tho, to see these in situ, with a rather cryptic error.

I can't trace the cause of the error as the tracebook doesn't show me where mdbook chokes, but I see the same issue when I try to run a build without any of the new diagrams and text I've added.

Can you let me know if this builds at your end, or if there are any changes you need?

```
~/C/t/scaphandre ❯❯❯ mdbook serve
2021-01-12 13:22:29 [ERROR] (mdbook::utils): Error: Summary parsing failed
2021-01-12 13:22:29 [ERROR] (mdbook::utils): 	Caused By: There was an error parsing the suffix chapters
2021-01-12 13:22:29 [ERROR] (mdbook::utils): 	Caused By: Error at line 21, column 1: Suffix chapters cannot be followed by a list
~/C/t/scaphandre ❯❯❯ mdbook clean
2021-01-12 13:22:43 [ERROR] (mdbook::utils): Error: Summary parsing failed
2021-01-12 13:22:43 [ERROR] (mdbook::utils): 	Caused By: There was an error parsing the suffix chapters
2021-01-12 13:22:43 [ERROR] (mdbook::utils): 	Caused By: Error at line 21, column 1: Suffix chapters cannot be followed by a list
```

For anyone else following along, I've been working in figma at the link below for making these diagrams:

https://www.figma.com/file/MnmCxOZlUWgUiFD0JDN692/understanding-scaphandre?node-id=0%3A1

bpetit commented:
I'll check if it builds on my end and try to figure out what's wrong. I'll keep you in touch.

Thanks !

bpetit commented:
Hi ! it does compile on my end. (which is strange :/)
I look at this and merge soon ! Thanks again !

bpetit commented on docs_src/explanations/how-scaph-computes-per-process-power-consumption.md:
```diff
@@ -1,7 +1,53 @@
-As you can see with the [prometheus exporter reference](../references/exporter-prometheus.md), scaphandre exporters can provide process level power consumption metrics. This section will explain how it is done and how it may be improved in the future.
+Scaphandre is a tool that makes it possible to see the power being used by a single process on a computer.
+
+This sounds like a simple thing thing to be able to do, but it turns out to be surprisingly complicated. So having a good mental for how it works will make it understand when and how to use Scaphandre.
```
I'd suggest to say "it's a bit more subbtle" or any sort of expression like that, rather than "complicated". It seems to me that the complexity is a subjective notion and some people using scaphandre maybe well aware of processes and tasks scheduling.

bpetit commented on docs_src/explanations/how-scaph-computes-per-process-power-consumption.md:
```diff
@@ -1,7 +1,53 @@
+
+#### Timesharing of work
+
+As we know computers work on lots of different jobs at the same time - they work on one job for short interval of time, then another, and another and so one. You'll often see these [small intervals of time referred to as [_jiffies_][].
```
Is there a missing reference/link here ?

bpetit commented on docs_src/explanations/how-scaph-computes-per-process-power-consumption.md:
```diff
@@ -1,7 +1,53 @@
+
+It's possible without Scaphandre to understand how large a share of a machines' resources are being used by a given process, but if we want to understand the power used per process, we need to know how much power is being used by the machine in absolute terms.
 
+To do this, we need a sensor of some kind to track power usage by the machien itself. Some servers have these like with Intel's RAPL sensots,, making it possible to understand how much power is being used by CPUs, GPUs and so on.
```
A typo here: "machien"

bpetit commented on docs_src/explanations/how-scaph-computes-per-process-power-consumption.md:
```diff
@@ -1,7 +1,53 @@
+
+As long are able to read from sensors that can share how much power is being used by a server at a given moment, and how much of the time  being allocated to our processes during those moments, we can calculate these figures.
+
+This can work with virtualised environments, as long the guest virtual machine or guest container has access to readings provided by the host physical machine. Scaphandre supports exposing power info to virtual machines run using QEMU/KVM. (More on scaphandre in virtualised scenarios to come.).
```
We should maybe mark the difference between virtual machines and containers, as it is not the same challenge.

mrchrisadams replied:
I think I'll need a bit more input in explaining the different challenges here, as I don't think I understand the container side problems well enough to write about them in detail yet.

This is partly why i was using less specific terms like _virtualised environments_, or _host physical machine_.

I can make an issue to document the different challenges faced in when working with different forms of virtualisation. 👍 ?

bpetit replied:
If you want to, that would be great !

bpetit commented:
I've added some comments to be debated.

In addition to those, I wonder if what you wrote wouldn't fit better in a specific section, like "An introduction to measuring power consumption" as you described pretty well the overall concept and the current section about processes is more about the implementation. Wdyt ?

mrchrisadams commented:
Yeah, I agree about the it maybe fitting in it's own section. I'm in favour of splitting it out, the docs here are largely introducing concepts that you might refer to in sections in implemention.

I'll update this PR later this afternoon, with a new page, and the points you raised addressed 👍

mrchrisadams commented:
Hi @bpetit,

I found out the issue with mdbook - My version of mdbook was too old, I had been bitten [by this issue here from a while back](https://github.com/rust-lang/mdBook/issues/576). I've since updated and it means I can generate the mdbook pages now - huzzah!

I've also added another diagram for multiprocessors, and tidied up the copy based on your review.

If you look the same page as rendered by github, you can see that the images are pretty large - they're smaller in the generated mdbook, and TBH, I'm not sure what the best way to make them smaller is for _just_ github. I've added a link to how it looks in the green web foundation repo so you can see:

https://github.com/thegreenwebfoundation/scaphandre/blob/ca-docs-for-per-process-consumption/docs_src/explanations/how-scaph-computes-per-process-power-consumption.md

Anyway, let me know if you're happy with this.

bpetit commented:
Merged, thanks a lot !
